fix!: refactor connection manager to use a prioritised queue

achingbrain opened the pull request:
Refactors the connection manager.

1. Internally it uses a queue to control concurrency instead of dial tokens (e.g. `maxParallelDials`)
2. A second queue is used for each peer to prevent inividual peers from swamping the dial queue (e.g. `maxParallelDialsPerPeer`)
3. The auto dialler now also checks the minimum connection limit on peer discovery and peer disconnect events for fast reconnect
4. Auto dialled peers are sorted based on tag value so valuable peers should be re-dialled first
5. Auto dialled peers are dialled in parallel to prevent a slow peer locking up the auto-dial queue
6. A `getDialQueue` method has been added to libp2p to allow inspection of the dial queue though it needs exposing in the interface
7. The connection gater `denyDialMultiaddr` method now only takes a multiaddr
8. Fixes duplicate connections to fast peers
9. Fixes unhanded promise rejections #1587
10. Connection pruning now happens after `peer:connect` events not before - fixes #1565

BREAKING CHANGE: some connection manager options have changed - please see the upgrade guide for full details

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    // this has been renamed to `maxParallelDialsPerPeer`
+    maxDialsPerPeer: 20,
+
+    // these should now be ranges expressed as MultiaddrFilters
+    allow: [
+      '/ip4/0.0.0.0/tcp/123'
+    ],
+
+    // these should now be ranges expressed as MultiaddrFilters
+    deny: [
+      '/ip4/0.0.0.0/tcp/123'
+    ]
```
Can we get a "allow peer" that allows any appropriate transports given a specific peerId?

I was trying to do that the other day in kubo and realized i needed a full multiaddr. I was trying to grant full-access to/from my helia node and couldn't figure out how to do it within an hour or so and i gave up.

It would be great to be able to do that in js-libp2p

achingbrain replied:
I'm a little confused, access is allowed by default, it's only disallowed if a configured connection gater denys it or you're at `maxConnections` already.

The `allow` list above enables specific networks to breach connection limits - a DoS mitigation for when your node is being eclipsed.

Maybe you can open an issue on the helia repo with a repro case?

SgtPooki replied:
sure. I will re-open if I run into it.. essentially what I was trying to do was prefer my local node to others, and to allow connections to that one no matter what. it sounds like that may not be necessary

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
````diff
@@ -0,0 +1,117 @@
+})
+```
+
+**Before**
````
```suggestion
**After**
```

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    // is above minConnections
+    autoDialConcurrency: 25,
+
+    // a low value allows user-initiated dials to take priority over
+    // auto dials
+    autoDialPriority: 0,
```
Nice

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    // auto dials
+    autoDialPriority: 0,
+
+    // this was previously named `maxAddrsToDial`
+    maxPeerAddrsToDial: 20,
```
PerConnection attempt? Or total(then we banlist the peer as bad??)

Might be good to call out which one and how in jsdoc, or in the name if possible.

achingbrain replied:
Did you check [the jsdoc for this config option](https://github.com/libp2p/js-libp2p/blob/fix/refactor-connection-manager/src/connection-manager/index.ts#L79-L83)?  If it's not clear could you suggest an edit?

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    // this was previously named `maxAddrsToDial`
+    maxPeerAddrsToDial: 20,
+
+    // this was previously named `maxDialsPerPeer`
+    maxParallelDialsPerPeer: 20,
```
Just so i understand fully, this setting changes whether we can attempt to dial 2 or 20 multiaddrs at once when attempting to connect to a specific peer, correct?

Would 'maxParallelMultiAddrDialsPerPeer' be a more explicit name? Is there a way we can get that explicitness without the verbosity?

achingbrain replied:
We only dial multiaddrs so maybe having `MultiAddr` in the name is redundant?

achingbrain replied:
But yes, a setting of `2` with a peer that has `10` addresses means we'll dial all 10 addresses but only ever 2 at once.  This stops a peer with a lot of spam addresses blocking the dial queue.

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    // this was previously named `maxDialsPerPeer`
+    maxParallelDialsPerPeer: 20,
+
+    // these should now be ranges expressed as MultiaddrFilters
+    allow: [
+      new MultiaddrFilter('/ip4/192.168.1.1/ipcidr/32')
+    ],
+
```
As above, can i create a multiaddrFilter consisting only of '/p2p/peerId' ?

achingbrain replied:
No, the allow/deny lists are for allowing/denying classes of network to make connections - this happens before the peer id exchange so we don't know who the remote peer is at that point.

It's a DoS mitigation strategy to always allow connections from networks you know, control or trust, or disallow connections from networks that are attacking you.

To deny specific peers you can implement a [connection gater](https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md#configuring-connection-gater).

SgtPooki commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+
+const node = await createLibp2p({
+  connectionGater: {
+    denyDialMultiaddr: (multiaddr) => {
+      if (multiaddr.getPeerId() != null) {
+        // there is a peer id present in the multiaddr
+      }
```
👍

SgtPooki commented on src/connection-manager/auto-dial.ts:
```diff
@@ -0,0 +1,148 @@
+
+export interface AutoDialInit {
+  /**
+   * The minimum number of connections below which libp2p will start to dial
+   * peers from the peer book. (default: 0)
+   */
+  minConnections?: number
```
MinConnections: 0 is the same as saying, I'm allowed to have zero connections, don't autoDial, correct?

Might be useful to call out that setting this to zero disables it

achingbrain replied:
Turns out that default was wrong 🤭  - would much rather generate that info but not sure how.

I've updated [the jsdoc for this setting](https://github.com/libp2p/js-libp2p/blob/fix/refactor-connection-manager/src/connection-manager/index.ts#L39-L44) - if it's still not clear could you suggest an edit?

SgtPooki replied:
https://github.com/libp2p/js-libp2p/blob/4078082dac944c6feb64e9e45e5275d7f0bf6ae4/src/connection-manager/index.ts#L39-L44 looks much better to me, thanks :)

SgtPooki commented on src/connection-manager/auto-dial.ts:
```diff
@@ -0,0 +1,148 @@
+  /**
+   * Proactively tries to connect to known peers stored in the PeerStore.
+   * It will keep the number of connections below the upper limit and sort
+   * the peers to connect based on wether we know their keys and protocols.
```
```suggestion
   * the peers to connect based on whether we know their keys and protocols.
```

SgtPooki reviewed (commented):
Partial review

maschad commented on src/connection-manager/dial-queue.ts:
```diff
@@ -0,0 +1,498 @@
+    const filteredAddrs = resolvedAddresses.filter(addr => Boolean(this.transportManager.transportForMultiaddr(addr.multiaddr)))
+
+    // deduplicate addresses
+    const dedupedAddrs: Map<string, Address> = new Map()
```
Couldn't we do something like 

```suggestion
let dedupedMultiaddrs = [...new Set(filteredAddrs.map(addr => { 
	addr.isCertified = addr.isCertified || addr.isCertified || false
})).values()]
```

to avoid the null checks per item

achingbrain replied:
`Address`es are not `Multiaddr`s - they have an `isCertified` field so we can't just stringify them into a set to dedupe, instead we use the keys of a map to dedupe but using the `Address` as map values lets us preserve the `isCertified` field.

maschad commented on src/connection-manager/dial-queue.ts:
```diff
@@ -0,0 +1,498 @@
+      const peerDialQueue = new PQueue({
+        concurrency: this.maxParallelDialsPerPeer
+      })
+      peerDialQueue.on('error', (err) => {
```
Should we abort dials here?

achingbrain replied:
No, because other dials in the queue may succeed - the error event may be emitted if an individual dial fails.

maschad commented on src/connection-manager/index.ts:
```diff
@@ -73,137 +90,175 @@ export interface ConnectionManagerConfig {
      * Map of connections per peer
      */
-    this.connections = new Map()
+    this.connections = new PeerMap()
```
Didn't know this existed! Nice!

maschad reviewed (changes requested):
Great work on this @achingbrain will definitely improve with connection management overall. I made some comments and suggestions. 

I think these are a lot of changes though and perhaps we should split this into some smaller PRs to make reviewing a bit easier and the changes more manageable. I would suggest separate PRs for:
- https://github.com/libp2p/js-libp2p/issues/1510
- https://github.com/libp2p/js-libp2p/issues/1630
- `denyDialMultiaddr` method now only takes a `multiaddr` and also the other interfaces updates required with it.

achingbrain commented on src/connection-manager/auto-dial.ts:
```diff
@@ -0,0 +1,140 @@
+    })
+
+    // shuffle the peers
+    peers = peers.sort(() => Math.random() > 0.5 ? 1 : -1)
```
So we don't dial the same peers every time.

Sorting is stable so if we start with a random order, then sort by a given field, where a subset of the peers have the same value for the given field they'll still be randomly sorted within that subset.

achingbrain commented on doc/migrations/v0.43-v0.44.md:
```diff
@@ -0,0 +1,117 @@
+    maxAddrsToDial: 20,
+
+    // this has been renamed to `maxParallelDialsPerPeer`
+    maxDialsPerPeer: 20,
```
Done!

achingbrain commented:
I've removed the attempt at resolving the double-peer discovery bug - I think it needs a bit more thought, basically the logic to determine whether to emit the event is complicated and will be vastly simplified by having an [atomic peer store](https://github.com/libp2p/js-libp2p-interfaces/pull/368).

The auto dial interval has also returned - this is because I found an edge case where we could still end up with no peers if all auto-dials failed - if network connection is lost, for example.

I've reverted the `MultiaddrFilter` change for allow/deny lists too - with a bit more thought it would be better to have the connection gater control allow/deny behaviour, then we aren't splitting responsibility for that between the connection manager and the connection gater.

achingbrain commented on doc/CONFIGURATION.md:
```diff
@@ -600,14 +600,7 @@ const node = await createLibp2p({
   connectionManager: {
     maxConnections: Infinity,
-    minConnections: 0,
-    pollInterval: 2000,
```
Done

achingbrain commented on doc/CONFIGURATION.md:
```diff
@@ -600,14 +600,7 @@ const node = await createLibp2p({
-    minConnections: 0,
-    pollInterval: 2000,
-    // The below values will only be taken into account when Metrics are enabled
-    maxData: Infinity,
```
Done

maschad reviewed (approved):
LGTM

achingbrain commented:
Please hold off on merging until https://github.com/jacobheun/any-signal/pull/25 is merged
